### PR TITLE
Add GitHub tokens pool support

### PR DESCRIPTION
**Pull request by lukaszgryglicki**
This adds support for multiple GitHub tokens for Perceval GitHub backend.

This is backward compatible.

You're specifying token as a string. If your token is just a single 40-hex-digit token then tokens pool will just have that single token. If your token is a comma-separated list of 40-hex-digit tokens that all tokens will be used when performing GitHub API calls.

Token with most API points will be used for each request. Checking for the best token happens every time you use 10% of the current token's remaining API points.

Changes are as small as possible and as compatible as possible, you can use Perceval GitHub without any changes at all, it will only be more immune to API points limit when providing more API tokens.

### CODE REVIEW

**Comment by lukaszgryglicki** [comment]
cc @valeriocos @jgbarah - I know that some test coverage fails, but please take an initial review - if that looks about OK, I'll try to update coverage tomorrow.

BTW: [json2hat tool](https://github.com/cncf/json2hat) was moved to its own repository and updated.
- Added support for fetching affiliations JSON from remote source (fallback when no local source is available)
- Added support for specifying both local and remote JSON source via environment variables
- Added support for setting sorting hat database parameters eiter via full connect string `SH_DSN` or by setting individual params via `SH_*` variables (or using default ones)
- Fixed bug when trying to update profile country code to the non-existing one (which caused foreign key violation and wasn't detected before because affiliations JSON was 4 times smaller)
- Fixed bug with company name collision when attempting to insert company name differing obly by lower/upper case from existing company (database company names are case in-sensitive in sorting hat mariadb database).
- Packaged in docker container
- Battle tested with real MariaDB managed database (AWS) and remote-only [github_users.json](
https://github.com/cncf/gitdm/blob/master/github_users.json?raw=true) (40+ Mb) from [cncf/gitdm](https://github.com/cncf/gitdm)

**Comment by lukaszgryglicki** [comment]
in the meantime can anybody point me to some information about how to run tests locally? I want to update them to pass with multiple tokens support.

**Comment by valeriocos** [comment]
HI @lukaszgryglicki overall the PR looks OK (I like the idea of having a pool of tokens). It could be nice to have the pool generalized to apply the same strategy to the other clients. It shouldn't be difficult to have some methods (e.g., `_choose_best_api_token`) in the `client.py` that can be redefined in the single clients like the `GitHubClient`.
What do you think @sduenas ?

**Comment by lukaszgryglicki** [comment]
I think this will be more complex when trying to make it generic - not sure if all other potential backends can behave similarly? Any chance to get it merged? Can you let me know how can I run test suite locally - so I will try to update test coverage as well?

**Comment by valeriocos** [comment]
> I think ... it generic - not sure if all other potential backends can behave similarly?

ok, it sounds reasonable.

> Any chance to get it merged?

I guess so, but let's wait for @sduenas opinion.

> Can you let me know how ... test coverage as well?

To run the tests locally, you should execute the file [run_tests]( https://github.com/chaoss/grimoirelab-perceval/blob/master/tests/run_tests.py). If you want to run just the tests for a given backend, you should execute the corresponding test file, for instance [test_github](/home/slimbook/Escritorio/sources/perceval/tests/test_github.py).

WRT test coverage, you can use the [coverage package](https://pypi.org/project/coverage/) in this way:
- `coverage run run_tests.py` to generate the `.coverage` file
- `coverage report -m` to list the uncovered lines

**Comment by lukaszgryglicki** [comment]
Thanks a lot, I'll return to test coverage when I have time and I'll update this PR and let you know.

**Comment by valeriocos** [comment]
you're welcome @lukaszgryglicki

**Comment by valeriocos** [comment]
I use the feature provided by pycharm (you have just to right click on the method and then run). However you can probably do it from [commandline](https://stackoverflow.com/questions/15971735/running-single-test-from-unittest-testcase-via-command-line) 

AFAIK we don't have testing documentation for `Perceval`

**Comment by lukaszgryglicki** [comment]
OK will struggle a bit. I'm developing, running and testing on a remote SSH server, and testi g full backend `GitHub` takes quite a long time:
```
run --source=perceval test_github.py
EEEEEEEEEEEE...EEEEEEEEEEEEE
```

**Comment by lukaszgryglicki** [comment]
Not really, output is scary:
```
======================================================================
ERROR: test_init (__main__.TestGitHubClient)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/root/.local/lib/python3.6/site-packages/urllib3/contrib/pyopenssl.py", line 453, in wrap_socket
    cnx.do_handshake()
  File "/usr/lib/python3/dist-packages/OpenSSL/SSL.py", line 1808, in do_handshake
    self._raise_ssl_error(self._ssl, result)
  File "/usr/lib/python3/dist-packages/OpenSSL/SSL.py", line 1540, in _raise_ssl_error
    raise SysCallError(errno, errorcode.get(errno))
OpenSSL.SSL.SysCallError: (32, 'EPIPE')

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/root/.local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 600, in urlopen
    chunked=chunked)
  File "/root/.local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 343, in _make_request
    self._validate_conn(conn)
  File "/root/.local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 839, in _validate_conn
    conn.connect()
  File "/root/.local/lib/python3.6/site-packages/urllib3/connection.py", line 344, in connect
    ssl_context=context)
  File "/root/.local/lib/python3.6/site-packages/urllib3/util/ssl_.py", line 344, in ssl_wrap_socket
    return context.wrap_socket(sock, server_hostname=server_hostname)
  File "/root/.local/lib/python3.6/site-packages/urllib3/contrib/pyopenssl.py", line 459, in wrap_socket
    raise ssl.SSLError('bad handshake: %r' % e)
ssl.SSLError: ("bad handshake: SysCallError(32, 'EPIPE')",)

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/requests/adapters.py", line 440, in send
    timeout=timeout
  File "/root/.local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 667, in urlopen
    **response_kw)
  File "/root/.local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 667, in urlopen
    **response_kw)
  File "/root/.local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 667, in urlopen
    **response_kw)
  [Previous line repeated 2 more times]
  File "/root/.local/lib/python3.6/site-packages/urllib3/connectionpool.py", line 638, in urlopen
    _stacktrace=sys.exc_info()[2])
  File "/root/.local/lib/python3.6/site-packages/urllib3/util/retry.py", line 398, in increment
    raise MaxRetryError(_pool, url, error or ResponseError(cause))
urllib3.exceptions.MaxRetryError: HTTPSConnectionPool(host='api.github.com', port=443): Max retries exceeded with url: /rate_limit (Caused by SSLError(SSLError("bad handshake: SysCallError(32, 'EPIPE')",),))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/dist-packages/httpretty/core.py", line 1055, in wrapper
    return test(*args, **kw)
  File "tests/test_github.py", line 1792, in test_init
    client = GitHubClient('zhquan_example', 'repo', 'aaa,bbb')
  File "/usr/local/lib/python3.6/dist-packages/perceval/backends/core/github.py", line 497, in __init__
    self._init_rate_limit()
  File "/usr/local/lib/python3.6/dist-packages/perceval/backends/core/github.py", line 785, in _init_rate_limit
    self._choose_best_api_token()
  File "/usr/local/lib/python3.6/dist-packages/perceval/backends/core/github.py", line 729, in _choose_best_api_token
    hdrs[idx] = super().fetch(url).headers
  File "/usr/local/lib/python3.6/dist-packages/perceval/client.py", line 131, in fetch
    response = self._fetch_from_remote(url, payload, headers, method, stream, verify)
  File "/usr/local/lib/python3.6/dist-packages/perceval/client.py", line 162, in _fetch_from_remote
    response = self.session.get(url, params=payload, headers=headers, stream=stream, verify=verify)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 533, in get
    return self.request('GET', url, **kwargs)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 520, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/lib/python3/dist-packages/requests/sessions.py", line 630, in send
    r = adapter.send(request, **kwargs)
  File "/usr/lib/python3/dist-packages/requests/adapters.py", line 506, in send
    raise SSLError(e, request=request)
requests.exceptions.SSLError: HTTPSConnectionPool(host='api.github.com', port=443): Max retries exceeded with url: /rate_limit (Caused by SSLError(SSLError("bad handshake: SysCallError(32, 'EPIPE')",),))

----------------------------------------------------------------------
Ran 1 test in 30.100s

FAILED (errors=1)
```

**Comment by lukaszgryglicki** [comment]
@sduenas @valeriocos test coverage is ready, please review. Would be great to have it merged. Battle-tested this on `cncf/devstats` repo with no token provided, with an empty token, with a wrong token, with a single correct token and multiple tokens (all wrong, wrong and good ones and all good). Added edge cases tests. Coverage is also OK.

File: perceval/backends/core/github.py [hunk 1]
```diff
@@ -704,19 +719,78 @@ def fetch_items(self, path, payload):
...
+                if self.rate_limit_header in hdr:
+                    rls[idx] = int(hdr[self.rate_limit_header])
+                else:
+                    rls[idx] = 0
+            max_idx = rls.index(max(rls))
+            self.last_checked = self.rate_limit
+            logger.debug("Remaining API points: {}, choosen index: {}".format(rls, max_idx))
+
+        # If we have any tokens use best of them
+        if len(self.token) > 0:
```
Comment (valeriocos) [inline perceval/backends/core/github.py]: The method `_need_check_tokens` returns False if there is one token, thus this condition can be probably removed and the nested statements can be included in the previous if statement, what do you think @lukaszgryglicki ?

File: perceval/backends/core/github.py [hunk 2]
```diff
@@ -704,19 +719,78 @@ def fetch_items(self, path, payload):
...
+
+        if len(self.token) <= 1:
+            return False
+        if self.rate_limit is None:
+            return True
+        if self.last_checked is None:
+            self.last_checked = self.rate_limit
+            return True
+        ratio = float(self.rate_limit) / float(self.last_checked)
+        if ratio < 0.9:
```
Comment (valeriocos) [inline perceval/backends/core/github.py]: it shouldn't be `ratio < 0.1` ?

File: perceval/backends/core/github.py [hunk 3]
```diff
@@ -68,7 +69,8 @@ class GitHub(Backend):
 
     :param owner: GitHub owner
     :param repository: GitHub repository from the owner
-    :param api_token: GitHub auth token to access the API
+    :param api_token: GitHub auth token(s) to access the API.
+        can be single token or comma separated list of tokens
```
Comment (valeriocos) [inline perceval/backends/core/github.py]: you could use the same description you wrote for the param `token` in `GitHubClient`

File: perceval/backends/core/github.py [hunk 4]
```diff
@@ -704,19 +719,78 @@ def fetch_items(self, path, payload):
                 items = response.text
                 logger.debug("Page: %i/%i" % (page, last_page))
 
+    def _choose_best_api_token(self):
+        """Check all API tokles defined and choose one with most remaining API points"""
+
+        url = urijoin(self.base_url, "rate_limit")
+        max_idx = 0
+        # If only one token given
```
Comment (valeriocos) [inline perceval/backends/core/github.py]: It seems that the comment is not consistent with the code below

File: perceval/backends/core/github.py (hunk 1 above)
Comment (lukaszgryglicki) [reply perceval/backends/core/github.py]: `_choose_best_api_token` is also called by `_init_rate_limit` so that `if` is needed.

File: perceval/backends/core/github.py [hunk 5]
```diff
@@ -704,19 +719,78 @@ def fetch_items(self, path, payload):
...
+            self.current_token = self.token[max_idx]
+            self.session.headers.update({'Authorization': 'token ' + self.current_token})
+
+    def _need_check_tokens(self):
+        """We need to consider changing token when last rate limit failed or allow using 10% of current token"""
+        """Only need to check rate again when used >10% of the last remaining rate"""
+
+        if len(self.token) <= 1:
+            return False
+        if self.rate_limit is None:
```
[the remark this answers is not in the document]
Comment (lukaszgryglicki) [reply perceval/backends/core/github.py]: I can even be in a single `if`, I'll change this.

File: perceval/backends/core/github.py (hunk 2 above)
Comment (lukaszgryglicki) [reply perceval/backends/core/github.py]: No, 0.9 is OK. It means that we're not checking for a better token until we use 10% of current token's remaining points.

**Comment by valeriocos** [comment]
please @lukaszgryglicki squash the commits. Then it can be merged if @sduenas doesn't have any comment

**Comment by lukaszgryglicki** [comment]
OK will do.

**Comment by lukaszgryglicki** [comment]
created a new PR with a single commit: https://github.com/chaoss/grimoirelab-perceval/pull/477
